### PR TITLE
feat:Adapt TongYi LLM speech transcription by Spring AI API

### PR DESCRIPTION
**Pull request by n3A87**
impl #3703

### CODE REVIEW

**Comment by ruansheng8** [comment]
Please run `mvn clean validate` first

**Review by yuluo-yx — CHANGES_REQUESTED** [review]
Good job, pls go ahead.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/TongYiAutoConfiguration.java [hunk 1]
```diff
@@ -18,15 +18,18 @@
 
 import java.util.Objects;
 
-import com.alibaba.cloud.ai.tongyi.audio.TongYiAudioSpeechClient;
-import com.alibaba.cloud.ai.tongyi.audio.TongYiAudioSpeechProperties;
+import com.alibaba.cloud.ai.tongyi.audio.speech.TongYiAudioSpeechClient;
+import com.alibaba.cloud.ai.tongyi.audio.speech.TongYiAudioSpeechProperties;
+import com.alibaba.cloud.ai.tongyi.audio.transcription.TongYiAudioTranscriptionClient;
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/TongYiAutoConfiguration.java:23]: And did you forget to add the client code, I don't see it in your pr.

you can exec `git status` to check your change code files.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/chat/TongYiChatOptionsTests.java [hunk 2]
```diff
@@ -37,14 +37,14 @@ class TongYiChatOptionsTests {
 	public void testChatOptions() {
 
 		Generation mockClient = Mockito.mock(Generation.class);
-		Constants.apiKey = "test";
+		Constants.apiKey = "sk-xxx";
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/chat/TongYiChatOptionsTests.java]: pls restore this changes, this change is not directly related to this pr.

we should maintain the singularity of pr

**Review by yuluo-yx — CHANGES_REQUESTED** [review]
Good job. pls go ahead

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/README.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+使用 curl 工具对接口发起请求：
+```shell
+$ curl -X GET "http://localhost:8080/ai/audio/transcription?audioUrls=url1&audioUrls=url2"
+
+# Response:
+D:\Code\spring-cloud-alibaba\05-13-20-47-08.txt
+D:\Code\spring-cloud-alibaba\05-13-20-47-09.txt
+```
+返回参数为保存到当前根路径下的音频转录文本文件的路径。
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/README.md]: Add blank line.

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	@Override
+	public String audioTranscription(List<String> audioUrls) {
+		AudioTranscriptionResponse res = audioTranscriptionClient.call(audioUrls);
+		List<TranscriptionTaskResult> taskResultList = res.getTranscriptionList();
+		return save(taskResultList);
+	}
+
+	private String save(List<TranscriptionTaskResult> taskResultList) {
+		String currentPath = System.getProperty("user.dir");
+
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java]: Delete this blank line.

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+						int bytesRead;
+
+						while ((bytesRead = in.read(dataBuffer, 0, 1024)) != -1) {
+
+							fileOutputStream.write(dataBuffer, 0, bytesRead);
+
+						}
+
+						logger.info("文件下载成功：{}", fileName);
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java]: pls use en.

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+						System.out.println();
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java]: you can use `\n`, avoid use out.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+import com.alibaba.dashscope.audio.asr.transcription.TranscriptionParam;
+import com.alibaba.dashscope.audio.asr.transcription.TranscriptionQueryParam;
+import com.alibaba.dashscope.audio.asr.transcription.TranscriptionResult;
+import com.alibaba.dashscope.audio.asr.transcription.TranscriptionTaskResult;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.util.Assert;
+
+
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java]: pls delete this blank line. retain a.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+	public AudioTranscriptionResponse call(AudioTranscriptionPrompt prompt) {
+		var param = merge(prompt.getOptions());
+		List<String> urls = prompt.getAudioUrl().getfileUrls();
+		TranscriptionParam transcriptionParam = toTranscriptionParam(param);
+		transcriptionParam.setFileUrls(urls);
+		logger.info(transcriptionParam.toString());
+
+		List<TranscriptionTaskResult> taskResultList = null;
+		try {
+			// 提交转写请求
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java]: pls use en.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java [hunk 9]
```diff
@@ -0,0 +1,172 @@
...
+			TranscriptionResult result = transcription.asyncCall(transcriptionParam);
+			// 等待转写完成
+			result = transcription.wait(TranscriptionQueryParam
+					.FromTranscriptionParam(transcriptionParam, result.getTaskId()));
+			// 获取转写结果
+			taskResultList = result.getResults();
+			return new AudioTranscriptionResponse(taskResultList);
+		}
+		catch (Exception e) {
+			System.out.println("error: " + e);
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java]: Should use `throw`, not `System.out.println()`

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionOptions.java [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import com.alibaba.cloud.ai.tongyi.audio.ParaformerModels;
+
+import org.springframework.ai.model.ModelOptions;
+
+
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionOptions.java]: ditto.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+import com.alibaba.dashscope.common.TaskStatus;
+import com.alibaba.dashscope.exception.ApiException;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.annotations.SerializedName;
+
+import org.springframework.ai.model.ModelResult;
+import org.springframework.ai.model.ResultMetadata;
+
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java]: ditto.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+
+
+/**
+ * @author: xYLiu
+ * @date: 2024/5/11 23:34
+ */
+
+public class AudioTranscription implements ModelResult<Object> {
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java]: pls remove to `metadata` dir.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+	public String toString() {
+		return "TranscriptionResult(requestId=" + this.getRequestId() + ", output="
+				+ this.getOutput() + ", usage=" + this.getUsage() + ", taskStatus="
+				+ this.getTaskStatus() + ", setTaskId=" + this.getsetTaskId()
+				+ ", results=" + this.getResults() + ", metrics=" + this.getMetrics()
+				+ ")";
+	}
+
+	@Override
+	public boolean equals(Object o) {
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java]: maybe wo not need `equals` method.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscriptionResponse.java [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.alibaba.cloud.ai.tongyi.audio.transcription.api;
+import java.util.List;
+import java.util.Objects;
+
+import com.alibaba.dashscope.audio.asr.transcription.TranscriptionTaskResult;
+
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscriptionResponse.java:20]: ditto.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioUrl.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * @author: xYLiu
+ * @date: 2024/5/10
+ */
+
+public class AudioUrl {
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioUrl.java]: We can delete this class file. imo, We not need.

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/TranscriptionClient.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+ * limitations under the License.
+ */
+
+package com.alibaba.cloud.ai.tongyi.audio.transcription.api;
+
+import java.util.Arrays;
+import java.util.List;
+
+@FunctionalInterface
+public interface TranscriptionClient {
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/TranscriptionClient.java]: maybe not need to define this interface, can use `modelclient` ，such as
 `https://github.com/spring-projects/spring-ai/blob/main/models/spring-ai-openai/src/main/java/org/springframework/ai/openai/OpenAiAudioTranscriptionClient.java`

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.when;
+
+/**
+ * @author: xYLiu
+ * @date: 2024/5/12
+ */
+
+public class TongYiAudioTranscriptionClientTest {
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java]: test class shoule not `public`, is `default`

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+		Transcription mockClient = Mockito.mock(Transcription.class);
+
+		TranscriptionResult expectedResult = new TranscriptionResult();
+		expectedResult.setResults(new ArrayList<>());
+		when(mockClient.asyncCall(any())).thenReturn(expectedResult);
+
+		var options = new TongYiAudioTranscriptionOptions();
+		var client = new TongYiAudioTranscriptionClient(options, mockClient);
+
+		// 假设AudioTranscriptionPrompt需要一个构造函数或工厂方法来创建
```
Comment (yuluo-yx) [inline spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java]: pls use en

**Review by yuluo-yx — CHANGES_REQUESTED** [review]

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/README.md (hunk 3 above)
Comment (n3A87) [reply spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/README.md]: reply done

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java (hunk 5 above)
Comment (n3A87) [reply spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java]: reply done

File: spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java (hunk 6 above)
Comment (n3A87) [reply spring-cloud-alibaba-examples/spring-cloud-ai-example/src/main/java/com/alibaba/cloud/ai/example/tongyi/service/impl/audio/transcription/TongYiAudioTranscriptionServiceImpl.java]: reply done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java (hunk 7 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java]: reply done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java (hunk 8 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java]: reply done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java (hunk 9 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClient.java]: reply done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java (hunk 12 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java]: done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java (hunk 13 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscription.java]: done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscriptionResponse.java (hunk 14 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioTranscriptionResponse.java:20]: done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java (hunk 17 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java]: done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java (hunk 18 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/test/java/com/alibaba/cloud/ai/tongyi/audio/transcription/TongYiAudioTranscriptionClientTest.java]: done

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioUrl.java (hunk 15 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/AudioUrl.java]: undone

File: spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/TranscriptionClient.java (hunk 16 above)
Comment (n3A87) [reply spring-cloud-alibaba-starters/spring-cloud-starter-alibaba-ai/src/main/java/com/alibaba/cloud/ai/tongyi/audio/transcription/api/TranscriptionClient.java]: undone

**Review by yuluo-yx — APPROVED** [review]
lgtm. tks @n3A87
